pkg/providers/amifamily/bootstrap/nodeadm.go

```diff
@@ -0,0 +1,142 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	admapi "github.com/awslabs/amazon-eks-ami/nodeadm/api"
+	admv1alpha1 "github.com/awslabs/amazon-eks-ami/nodeadm/api/v1alpha1"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily/bootstrap/mime"
+)
+
+type Nodeadm struct {
+	Options
+}
+
+func (n Nodeadm) Script() (string, error) {
+	nodeConfigYAML, err := n.getNodeConfigYAML()
+	if err != nil {
+		return "", fmt.Errorf("generating NodeConfig, %w", err)
+	}
+	customEntries, err := n.parseUserData()
+	if err != nil {
+		return "", fmt.Errorf("parsing custom UserData, %w", err)
+	}
+	mimeArchive := mime.Archive(append([]mime.Entry{{
+		ContentType: mime.ContentTypeNodeConfig,
+		Content:     nodeConfigYAML,
+	}}, customEntries...))
+	userData, err := mimeArchive.Serialize()
+	if err != nil {
+		return "", err
+	}
+	return userData, nil
+}
+
+// getNodeConfigYAML returns the Karpenter generated NodeConfig YAML object serialized as a string
+func (n Nodeadm) getNodeConfigYAML() (string, error) {
+	config := &admv1alpha1.NodeConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       admapi.KindNodeConfig,
+			APIVersion: admv1alpha1.GroupVersion.String(),
+		},
+		Spec: admv1alpha1.NodeConfigSpec{
+			Cluster: admv1alpha1.ClusterDetails{
+				Name:              n.ClusterName,
+				APIServerEndpoint: n.ClusterEndpoint,
+				CIDR:              n.ClusterCIDR,
+			},
+		},
+	}
+	if lo.FromPtr(n.CABundle) != "" {
+		ca, err := base64.StdEncoding.DecodeString(*n.CABundle)
+		if err != nil {
+			return "", fmt.Errorf("decoding CABundle, %w", err)
+		}
+		config.Spec.Cluster.CertificateAuthority = ca
+	}
+	inlineConfig, err := n.generateInlineKubeletConfiguration()
+	if err != nil {
+		return "", err
+	}
+	config.Spec.Kubelet.Config = inlineConfig
+	if arg := n.nodeLabelArg(); arg != "" {
+		config.Spec.Kubelet.Flags = []string{arg}
+	}
+
+	// Convert to YAML at the end for improved legibility.
+	configYAML, err := yaml.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("# Karpenter Generated NodeConfig\n%s", string(configYAML)), nil
+}
+
+// generateInlineKubeletConfiguration returns a serialized form of the KubeletConfiguration specified by the Nodeadm
+// options, for use with nodeadm's NodeConfig struct.
+func (n Nodeadm) generateInlineKubeletConfiguration() (map[string]runtime.RawExtension, error) {
+	kubeConfigJSON, err := json.Marshal(n.KubeletConfig)
```

Comment:
I think it's reasonable to take this stance because it's tied to the KubeletConfiguration API, not anything `nodeadm` specific. I think our stance has been that the names should align with the upstream API naming anyway.

Reply:
Agreed. I think that it's fine. Just a callout that we need to watch out for.